incubator/hnc/internal/forest/namespaceobjects.go

```diff
@@ -5,76 +5,77 @@ import (
+	if len(ns.sourceObjects[gvk]) == 0 {
+		delete(ns.sourceObjects, gvk)
 	}
 }
 
-// GetOriginalObjects returns all original objects in the namespace.
-func (ns *Namespace) GetOriginalObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
+// GetSourceObjects returns all source objects in the namespace.
+func (ns *Namespace) GetSourceObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
 	o := []*unstructured.Unstructured{}
-	for _, obj := range ns.originalObjects[gvk] {
+	for _, obj := range ns.sourceObjects[gvk] {
 		o = append(o, obj)
 	}
 	return o
 }
 
-// GetNumOriginalObjects returns the total number of original objects of a specific GVK in the namespace.
-func (ns *Namespace) GetNumOriginalObjects(gvk schema.GroupVersionKind) int {
-	return len(ns.originalObjects[gvk])
+// GetNumSourceObjects returns the total number of source objects of a specific
+// GVK in the namespace.
+func (ns *Namespace) GetNumSourceObjects(gvk schema.GroupVersionKind) int {
+	return len(ns.sourceObjects[gvk])
 }
 
-// GetPropagatedObjects returns all original copies in the ancestors.
-func (ns *Namespace) GetPropagatedObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
+// GetAncestorSourceObjects returns all source objects in the ancestors.
+func (ns *Namespace) GetAncestorSourceObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
 	o := []*unstructured.Unstructured{}
 	ans := ns.AncestryNames()
 	for _, n := range ans {
 		// Exclude the original objects in this namespace
 		if n == ns.name {
 			continue
 		}
-		o = append(o, ns.forest.Get(n).GetOriginalObjects(gvk)...)
+		o = append(o, ns.forest.Get(n).GetSourceObjects(gvk)...)
 	}
 	return o
 }
 
-// GetPropagatingObjects returns all the source objects to be propagated into the
-// descendants of this namespace.
+// GetAncestorAndLocalSourceObjects returns all the source objects to be
+// propagated into the descendants of this namespace.
 // TODO update this function to reflect the changes introduced by the future HNC
 //  exceptions implementation.
-func (ns *Namespace) GetPropagatingObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
-	return append(ns.GetPropagatedObjects(gvk), ns.GetOriginalObjects(gvk)...)
+func (ns *Namespace) GetAncestorAndLocalSourceObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
+	return append(ns.GetAncestorSourceObjects(gvk), ns.GetSourceObjects(gvk)...)
 }
 
-// GetSource returns the original copy in the ancestors if it exists.
-// Otherwise, return nil.
-func (ns *Namespace) GetSource(gvk schema.GroupVersionKind, name string) *unstructured.Unstructured {
-	pos := ns.GetPropagatedObjects(gvk)
+// GetAncestorSourceObject returns the source object in the ancestors if it
+// exists. Otherwise, return nil.
+func (ns *Namespace) GetAncestorSourceObject(gvk schema.GroupVersionKind, name string) *unstructured.Unstructured {
+	pos := ns.GetAncestorSourceObjects(gvk)
```

Comment:
I don't think this is correct anymore because it doesn't account for the fact that the most recent source object with the same name might not have been propagated (and this will be even more important with exceptions). This should probably return a list, starting with any object in the most recent ancestor.

You could even combine this with `GetAncestorSourceObjects` (since this is no longer returning a single object anymore). If `name` is empty, it returns everything; otherwise it _only_ matches objects with the same name.

Reply:
True, I updated it with a list but still starting with the top ancestor since the reconciler behavior is to propagate the top "propagatable" ancestor (with no finalizer).